### PR TITLE
Public Auth API

### PR DESCRIPTION
**Pull request by tomchristie**
Refs #731

Starting this PR with the proposed documentation. Implementation will require:

- [x] Switching the auth method name to `.auth_flow`.
- [x] Implementing a `Request.aread()` method and `Request.content` property.
- [x] Implementing `.requires_request_body` functionality on auth flows.
- [x] Expose `httpx.Auth` as public API.
- [x] Drop `AuthFlow` in favor of explicit type, in order to help guide users away from importing bits of `httpx` that are outside public API.

### CODE REVIEW

**Comment by florimondmanca** [comment]
Cc'ing @lepture relative to https://github.com/lepture/authlib/issues/178#issuecomment-571153749 — from the docs proposed here, would this fit the use case for `authlib`?

File: tests/models/test_requests.py [hunk 1]
```diff
@@ -21,19 +21,38 @@ def test_content_length_header():
...
 
     assert request.headers["Content-Type"] == "application/json"
-    assert content == b'{"test": 123}'
+    assert request.content == b'{"test": 123}'
+
+
+@pytest.mark.asyncio
+async def test_read_and_stream_data():
+    # Ensure a request may still be streamed if it has been read.
+    #  Needed for cases such as authentication classes that read the request body.
```
Comment (florimondmanca) [inline tests/models/test_requests.py]: Nit:

```suggestion
    # Needed for cases such as authentication classes that read the request body.
```

Same on the comment in the other test case below.

**Comment by tomchristie** [comment]
> I'm really happy with how the auth API turned out

I think you made a great call on the generator approach. 👍

It's also kinda nice that we've got a `callable[request] -> request` available for the simple cases, and a more powerful custom class + `.auth_flow` for multi-request flows.

**Comment by lepture** [comment]
yes, that would work.

**Pull request merged**